src/v/archival/ntp_archiver_service.cc

```diff
@@ -0,0 +1,290 @@
+    if (segment) {
+        // match segment with metadata
+        if (!validate_metadata(segment, meta)) {
+            co_return false;
+        }
+        auto stream = segment->offset_data_stream(
+          meta.base_offset, ss::default_priority_class());
+        auto s3path = _remote.get_remote_segment_path(sname);
+        vlog(archival_log.trace, "Uploading segment \"{}\" to S3", s3path());
+        try {
+            co_await client.put_object(
+              _bucket,
+              s3::object_key(s3path()),
+              segment->size_bytes(),
+              std::move(stream));
+            vlog(
+              archival_log.trace, "Completed segment \"{}\" to S3", s3path());
+            _remote.add(sname, meta);
+        } catch (...) {
+            vlog(
+              archival_log.error,
+              "Failed to upload {} to S3. Reason: {}",
+              sname,
+              std::current_exception());
+            co_return false;
+        }
+    } else {
+        vlog(archival_log.error, "Can't find segment {}", sname());
+    }
+    co_return true;
+}
+
+ss::future<ntp_archiver::batch_result> ntp_archiver::upload_next_candidate(
+  ss::semaphore& req_limit, storage::log_manager& lm) {
+    vlog(archival_log.trace, "Uploading next candidate called");
+    gate_guard guard{_gate};
+    // calculate candidate set
+    auto local = _policy->make_local_manifest(lm);
+    if (!local) {
+        vlog(archival_log.error, "Failed to generate local manifest");
+        co_return batch_result{};
+    }
+    auto candidates = local->difference(_remote);
+    auto num_candidates = std::min(
+      static_cast<ssize_t>(candidates.size()), req_limit.available_units());
+    if (num_candidates == 0) {
+        co_return batch_result{};
+    }
+    // NOTE: here consume API is used because we don't want the number of
+    // available semaphore units to change.
+    auto sem_units = ss::consume_units(req_limit, num_candidates);
+    std::vector<manifest::segment_map::value_type> upload_set;
+    std::copy_n(
+      candidates.begin(), num_candidates, std::back_inserter(upload_set));
+    vlog(archival_log.trace, "Upload {} elements", upload_set.size());
+    s3::client client(_client_conf);
+    batch_result result{};
+    auto fut = ss::parallel_for_each(
+      upload_set,
+      [this, &lm, &result](manifest::segment_map::value_type target) {
+          return upload_segment(std::move(target), lm).then([&result](bool ok) {
+              if (ok) {
+                  result.succeded++;
+              } else {
+                  result.failed++;
+              }
+          });
+      });
+    co_await std::move(fut);
+    if (num_candidates) {
+        vlog(archival_log.trace, "Completed S3 upload");
+        co_await client.shutdown();
+        co_await upload_manifest();
+        _last_upload_time = ss::lowres_clock::now();
+    }
+    co_return result;
+}
+
+ss::future<bool> ntp_archiver::delete_segment(
```

Comment:
do you think that in the future it may make sense to have deletion also be a policy? for example, a cluster may stream data into s3 and at any moment we can construct some consistent version of the cluster from the past. but if we wanted to protect against accidental deletions, for example, then maybe we'd want to also have a policy for garbage collection of data in the cloud?